### ISSUE TITLE
No editor command `WikiLinkFollow`.

### ISSUE THREAD

**Issue by PhilippFeO** [lervag/wiki.vim#279]
# Description

1. Typing `<CR>` in normal mode or `WikiLinkFollow` in command mode creates (markdown) Link.
2. Nothing happens, i.e. cursor descends one line (if available) or error message `E492: no editor command: WikiLinkFollow` appears.
3. To reproduce my issue, probably my full config is necessary. When I start Neovim with `nvim -u init-wiki.lua`, where `init-wiki.lua` only contains `wiki.vim` related options (s. below), everything works fine.

# Minimal working example

Wiki structure:

```
~/wiki.vim
└── index.md
```

Contents of `index.md`:

```
Lorem Ipsum
```

Contents of `init-wiki.lua` (basically my `wiki.vim` settings when using my config):

```lua
vim.g.wiki_root = '~/wiki.vim'
vim.g.wiki_filetypes = { 'md' }
vim.g.wiki_link_extension = '.md'
vim.g.wiki_link_target_type = 'md'
vim.g.wiki_global_load = 0
```

---

I searched the issues for potential conflicts with other plugins but there are none. Currently, I use packer, nvim-treesitter, vimtex, [frazrepo/vim-rainbow](https://github.com/frazrepo/vim-rainbow), coc.nvim, [honza/vim-snippets](https://github.com/honza/vim-snippets).

Disclaimer: I am relative new in the Vim/Neovim world, so I might lacking knowledge of internal mechanics or have beginner config flaws.

Edit: Any hint is welcome, even the answer that it is hard to detect/fix error without using my config (because you can't give everyone a personal cure).

**Comment by lervag** [comment]
This is very likely an issue with your configuration. It is currently hard to help, because I don't know much about your config. Do you host it on GitHub or similar; if so, feel free to share a link. If not, then you could also consider sharing your entire config here (e.g. a zip file).

> 3. To reproduce my issue, probably my full config is necessary. When I start Neovim with `nvim -u init-wiki.lua`, where `init-wiki.lua` only contains `wiki.vim` related options (s. below), everything works fine.

> Contents of `init-wiki.lua` (basically my `wiki.vim` settings when using my config):
> 
> ```lua
> vim.g.wiki_root = '~/wiki.vim'
> vim.g.wiki_filetypes = { 'md' }
> vim.g.wiki_link_extension = '.md'
> vim.g.wiki_link_target_type = 'md'
> vim.g.wiki_global_load = 0
> ```

You claim this should work. But how does this load wiki.vim? Is it loaded as a standard Vim/neovim package with the `:pack` feature?

Learning to build minimal example configurations is quite useful, as it makes it much easier to debug most kinds of problems. The following would be a standard minimal configuration (in vimscript; lua is very similar):

```vim
" test.vim
set runtimepath^=~/.local/plugged/wiki.vim
filetype plugin indent on
syntax enable

let g:wiki_root = '~/wiki.vim'
let g:wiki_filetypes = ['md']
let g:wiki_link_extension = '.md'
let g:wiki_link_target_type = 'md'
let g:wiki_global_load = 0
```

Of course, change the `runtimepath` to the correct path to wiki.vim on your system. Then you can use it like this: `nvim --clean -u test.vim`. Now doing `:WikiIndex` should bring you to `~/wiki.vim/index.md`.

> I searched the issues for potential conflicts with other plugins but there are none. Currently, I use packer, nvim-treesitter, vimtex, [frazrepo/vim-rainbow](https://github.com/frazrepo/vim-rainbow), coc.nvim, [honza/vim-snippets](https://github.com/honza/vim-snippets).

No, this is not a problem with conflicts. This is a problem where for some reason you are not properly loading wiki.vim. I would guess you are using lazy.nvim or packer and that you are somehow lazy loading wiki.vim. I found it necessary to use the `init` key for specifying config for wiki.vim with lazy.nvim.

> Disclaimer: I am relative new in the Vim/Neovim world, so I might lacking knowledge of internal mechanics or have beginner config flaws.

We all start out as beginners. Enjoy the journey! :)

**Comment by PhilippFeO** [comment]
> No, this is not a problem with conflicts. This is a problem where for some reason you are not properly loading wiki.vim. I would guess you are using lazy.nvim or packer and that you are somehow lazy loading wiki.vim.

This hint solved the issue! I changed (my plugin manager is packer.nvim) `use 'lervag/wiki.vim'` to `use { 'lervag/wiki.vim', opt = false, event = 'VimEnter' }` and now links are generated when hitting `<CR>` and I can follow them. I don't know why but both keys are necessary.

> I found it necessary to use the `init` key for specifying config for wiki.vim with lazy.nvim.

Can you tell me exactly, how you configured wiki.vim with lazy.nvim? I think about "switching" to [kickstart.nvim](https://github.com/nvim-lua/kickstart.nvim) which uses lazy.nvim. I've seen the `init` key but currently I don't know how to use it.

---

Although my issue is solved, I want to give you some answers/thoughts on your reply, in case you are curious where beginners might have problems.

> This is very likely an issue with your configuration. It is currently hard to help, because I don't know much about your config. Do you host it on GitHub or similar; if so, feel free to share a link. If not, then you could also consider sharing your entire config here (e.g. a zip file).

[My config](https://github.com/PhilippFeO/dotfiles/tree/master/nvim) is on GitHub but I am not very proud of it (one reason I think about switching to `kickstart.nvim`). It is your own responsibility if you decide inspecting it. :D

> You claim this should work. But how does this load wiki.vim? Is it loaded as a standard Vim/neovim package with the `:pack` feature?

I don't actively load wiki.vim in this scenario (as far as I know as a beginner). It just worked when using this file as my init.lua. In general, I added `use 'lervag/wiki.vim'` in my packer.nvim startup and then did `:PackerSync`, afterwards I didn't touch deliberately any loading mechanic.

> Learning to build minimal example configurations is quite useful, as it makes it much easier to debug most kinds of problems. The following would be a standard minimal configuration (in vimscript; lua is very similar):

To provide a minimal example was the reason I crafted the working init.lua. Then, I was puzzled because I didn't expect it to work.
To summarize your approach (and following next time): By setting `runtimepath` to wiki.vim and using `--clean -u test.vim`, you only include/use the wiki.vim plugin (and ignore other installed ones)?

> We all start out as beginners. Enjoy the journey! :)

Thank you, especially for the offer to check my config! 😊

**Comment by lervag** [comment]
> This hint solved the issue! I changed (my plugin manager is packer.nvim) `use 'lervag/wiki.vim'` to `use { 'lervag/wiki.vim', opt = false, event = 'VimEnter' }` and now links are generated when hitting `<CR>` and I can follow them. I don't know why but both keys are necessary.

Glad to hear it! I don't have any experience with packer. It's weird that the simple `use ...` did not work. But the important thing is that it works :)
> 
> > I found it necessary to use the `init` key for specifying config for wiki.vim with lazy.nvim.
> 
> Can you tell me exactly, how you configured wiki.vim with lazy.nvim? I think about "switching" to [kickstart.nvim](https://github.com/nvim-lua/kickstart.nvim) which uses lazy.nvim. I've seen the `init` key but currently I don't know how to use it.

Yes, see [here](https://github.com/lervag/dotnvim/blob/07b03b560fae3b094a5af7cfcb5c1be5026bb0d9/lua/lervag/init/packages.lua#L66-L125).

> I don't actively load wiki.vim in this scenario (as far as I know as a beginner). It just worked when using this file as my init.lua. In general, I added `use 'lervag/wiki.vim'` in my packer.nvim startup and then did `:PackerSync`, afterwards I didn't touch deliberately any loading mechanic.

Ah, ok. I believe the reason is that you are loading the wiki.vim configuration _after_ wiki.vim is itself loaded. You need to set the options _before_ wiki.vim is loaded. I think the change you did to the plugin loading works because it postpones loading of wiki.vim. But if you instead move the wiki.vim config so that it is loaded on startup, then you should not need the change that "fixed" things. I can look at your config to help you if you need help with that.

> > Learning to build minimal example configurations is quite useful, as it makes it much easier to debug most kinds of problems. The following would be a standard minimal configuration (in vimscript; lua is very similar):
> 
> To provide a minimal example was the reason I crafted the working init.lua. Then, I was puzzled because I didn't expect it to work. To summarize your approach (and following next time): By setting `runtimepath` to wiki.vim and using `--clean -u test.vim`, you only include/use the wiki.vim plugin (and ignore other installed ones)?

Exactly. The `--clean` option says to _not_ load plugins from your normal paths. When you do `nvim -u test.vim`, neovim still loads plugins that you put in your `~/.config/nvim/plugin` path and similar locations.

> > We all start out as beginners. Enjoy the journey! :)
> 
> Thank you, especially for the offer to check my config! blush

My pleasure :)

**Comment by PhilippFeO** [comment]
> Yes, see [here](https://github.com/lervag/dotnvim/blob/07b03b560fae3b094a5af7cfcb5c1be5026bb0d9/lua/lervag/init/packages.lua#L66-L125).

Nice, thank you for the example!
Personal question: Based on your name and the snow on your picture, I assumed you have Swedish or Norwegian roots but you are using German names for the names. Are you one of the lucky and happy people who grew up in an international environment? :D

> Ah, ok. I believe the reason is that you are loading the wiki.vim configuration _after_ wiki.vim is itself loaded. You need to set the options _before_ wiki.vim is loaded. I think the change you did to the plugin loading works because it postpones loading of wiki.vim. But if you instead move the wiki.vim config so that it is loaded on startup, then you should not need the change that "fixed" things. I can look at your config to help you if you need help with that.

I read the `:h startup` and as far as I can tell, there should be no messy loading queue. My wiki.vim config is in `/after/plugin` which should be loaded after `~/.local/nvim/site`/the plugins.
If you have more information or additional sources on the loading behavior, please let me know it! :)

I am very glad for you offer to check my config in tandem but I will decline it, mainly because of my thought about using kickstart.nvim. I also have to admit shyly that I feel honored that you would use your spare time in the form of checking my config, which is beyond providing/maintaining a plugin.

> Exactly. The `--clean` option says to _not_ load plugins from your normal paths. When you do `nvim -u test.vim`, neovim still loads plugins that you put in your `~/.config/nvim/plugin` path and similar locations.

Nice, I grasped the idea correctly!

**Comment by PhilippFeO** [comment]
> `plugin/wiki.vim` from wiki.vim will be loaded _earlier_ than all of your own `after/plugin/*.{vim,lua}` scripts. Thus, any option you define there will not apply to the things specified above.
> 
> So, to be clear: you should move your wiki.vim config into your `init.lua` or `init.vim` file or any file that is explicitly sourced by it. I think it is quite safe to say that you should do this for the configuration of _all_ plugins, unless you really know what you are doing.

I put the files where I set plugin specific options deliberately in `after/plugin/` because I thought it makes sense to set these options last, i. e. after the plugin is loaded because then the options are/plugin is guaranteed to be available.
Your answer sounds a bit like "set options before plugins are loaded". Do I get this right?

Just to be clear:
Although I set only markdown related options in my `wiki.lua`, which (kinda) worked because a `index.md` was generated, it interferes/might interfere with the loading behavior leading to missing editor commands like `WikiLinkFollow`?
Or do `let g:wiki_link_extension = '.md'` and `let g:wiki_link_target_type = 'md'` break the configured `WikiLinkFollow` command because it is setup to work with the `wiki` format and there is actually no editor command `WikiLinkFollow for md files`?

If so, I am curious why other, more complex plugins like treesitter, telescope and my LSP are loaded correctly. (Please don't contemplate this as a question directed to you, it is meant as a closing thought on an issue I have to investigate.)

**Comment by lervag** [comment]
> I put the files where I set plugin specific options deliberately in `after/plugin/` because I thought it makes sense to set these options last, i. e. after the plugin is loaded because then the options are/plugin is guaranteed to be available. Your answer sounds a bit like "set options before plugins are loaded". Do I get this right?

Yes, exactly.

> Just to be clear: Although I set only markdown related options in my `wiki.lua`, which (kinda) worked because a `index.md` was generated, it interferes/might interfere with the loading behavior leading to missing editor commands like `WikiLinkFollow`? …

I think the main reason it doesn't work is because the option `let g:wiki_filetypes = ['md']` is _only_ used during wiki.vim initialization. Since that happened _before_ `after/plugin/wiki.lua`, it means wiki.vim is not loaded for `.md` files.

> If so, I am curious why other, more complex plugins like treesitter, telescope and my LSP are loaded correctly. (Please don't contemplate this as a question directed to you, it is meant as a closing thought on an issue I have to investigate.)

Many of these are lua plugins that you initialize explicitly with `require("plugin").setup(...)`. But some of them are not. So for these: Most plugins expose functionality as commands or mappings. Configuration variables for these plugins should as a best practise be read _when needed_. And for many plugins, this is how it works. For wiki.vim, it is also the case for many options. Thus, you can change some options "live" and see the effect. However, wiki.vim exposes some functionality that is activated during initialization. Thus, if you want to change the behaviour by changing the options, then you must change the options before the plugin is loaded.

IMHO, you should not worry about the cost of defining options. The startuptime is mostly unaffected by `let ...` statements (or similar lua code). So, by ensuring that your plugin configuration is specified before a plugin is loaded you will not get into any trouble. If you, on the other hand, postpone plugin configuration, then you _may_ get into trouble with some plugins or some plugin features - such as in this very case we are discussing.

I hope this makes things slightly more clear! :)

**Comment by PhilippFeO** [comment]
> Yes, exactly.

I didn't expect to be able to access plugin variables before their loading procedure. :D 

> Many of these are lua plugins that you initialize explicitly with `require("plugin").setup(...)`. But some of them are not. So for these: Most plugins expose functionality as commands or mappings. Configuration variables for these plugins should as a best practise be read _when needed_. And for many plugins, this is how it works. For wiki.vim, it is also the case for many options. Thus, you can change some options "live" and see the effect. However, wiki.vim exposes some functionality that is activated during initialization. Thus, if you want to change the behaviour by changing the options, then you must change the options before the plugin is loaded.
>
> IMHO, you should not worry about the cost of defining options. The startuptime is mostly unaffected by `let ...` statements (or similar lua code). So, by ensuring that your plugin configuration is specified before a plugin is loaded you will not get into any trouble. If you, on the other hand, postpone plugin configuration, then you _may_ get into trouble with some plugins or some plugin features - such as in this very case we are discussing.
> 
> I hope this makes things slightly more clear! :)

It **absolutely** does! Currently, I contemplate startup time optimization as a topic for advanced users but since I want to reach this state (sometime), I will keep in mind!
By now, I have no open questions anymore and I know how to handle plugin configuration in the future! 🤩 It is not possible, but now I would close the issue a second time! :D

---

Although, it's not vimtex, but because we momentarily stay in touch: I really like vimtex! It makes working with LaTeX (more) convenient than I expected [^1]. I was also able to setup InverseSearch with Okular (for me as a beginner a success :D). Even though I only use basic features, it improves my (rudimentary) LaTeX workflow, so: Thank you!

[^1]: I basically expected TexStudio with some Vim motions but the embedding in the Vim/Neovim environment is just superior.

**Comment by lervag** [comment]
> It **absolutely** does!

Great, glad to hear it! Good luck on your continued journey with Vim/neovim! :)

> Although, it's not vimtex, but because we momentarily stay in touch: I really like vimtex! It makes working with LaTeX (more) convenient than I expected. I was also able to setup InverseSearch with Okular (for me as a beginner a success :D). Even though I only use basic features, it improves my (rudimentary) LaTeX workflow, so: Thank you!

Thank you for the kind words! I'm glad to hear my work is useful to you!
